### PR TITLE
[Model] Add torch.compile support for Ovis2.5 vision encoder

### PR DESCRIPTION
**Pull request by tianrengao**
# [Model] Add torch.compile support for Ovis2.5 vision encoder

## Purpose

Enable `torch.compile` for the Siglip2 vision encoder used in Ovis2.5 model to improve inference throughput.

Changes:
- Add `@support_torch_compile` decorator to `Siglip2EncoderLayer` with proper `dynamic_arg_dims` configuration
- Split `position_embeddings` tuple into separate `rotary_pos_emb_cos` and `rotary_pos_emb_sin` tensor arguments for `dynamic_arg_dims` compatibility (tuples are not supported by `dynamic_arg_dims`)
- Wrap layer initialization with `set_model_tag` for proper compile cache key generation
- Add `set_forward_context` wrapper in `ovis2_5.py` for compilation context

## Test Plan

```bash
# Start server with compile enabled
vllm serve AIDC-AI/Ovis2.5-2B --trust-remote-code --compilation-config='{"compile_mm_encoder":"true"}'

# Run benchmark
vllm bench serve --backend openai-chat --model AIDC-AI/Ovis2.5-2B --endpoint /v1/chat/completions --dataset-name hf --dataset-path lmarena-ai/VisionArena-Chat --hf-split train --num-prompts 1000
```

## Benchmark Result

**Model:** AIDC-AI/Ovis2.5-2B
**GPU:** NVIDIA H100 80GB
**Dataset:** lmarena-ai/VisionArena-Chat (1000 prompts)

| Metric | Eager | Compiled | Improvement |
|--------|-------|----------|-------------|
| Request throughput (req/s) | 32.01 | 37.74 | **+17.9%** |
| Output token throughput (tok/s) | 4,072 | 4,805 | **+18.0%** |
| Total token throughput (tok/s) | 7,092 | 8,365 | **+18.0%** |
| Mean TTFT (ms) | 17,372 | 17,788 | +2.4% |
| Mean TPOT (ms) | 99.84 | 50.05 | **-49.9%** |
| Median TPOT (ms) | 88.53 | 25.49 | **-71.2%** |
| Mean ITL (ms) | 99.81 | 51.87 | **-47.5%** |
| Median ITL (ms) | 57.48 | 23.11 | **-59.8%** |

**~18% throughput improvement** with torch.compile enabled for the vision encoder.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/model_executor/models/ovis2_5.py [hunk 1]
```diff
@@ -556,9 +558,10 @@ def _process_visual_input(
         )
 
         target_dtype = self.visual_tokenizer.dtype
-        visual_tokens = self.visual_tokenizer(
-            image_patches_flat.to(target_dtype), grid_thws
-        )
+        with set_forward_context(None, self.vllm_config):
```
Comment (Lucaskabela) [inline vllm/model_executor/models/ovis2_5.py]: set_forward context not needed anymore :)

File: vllm/model_executor/models/siglip2navit.py [hunk 2]
```diff
@@ -309,6 +313,16 @@ def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
...
+    dynamic_arg_dims={
+        "hidden_states": 0,
+        "cu_seqlens": 0,
+        "rotary_pos_emb_cos": 0,
+        "rotary_pos_emb_sin": 0,
+    },
+    enable_if=should_torch_compile_mm_encoder,
+    is_encoder=True,
+)
 class Siglip2EncoderLayer(nn.Module):
```
Comment (Lucaskabela) [inline vllm/model_executor/models/siglip2navit.py:326]: Can we apply this on the entire Siglip2Encoder? Or is there some incompatability with that?

File: vllm/model_executor/models/siglip2navit.py [hunk 3]
```diff
@@ -377,16 +394,20 @@ def __init__(
...
-                Siglip2EncoderLayer(
-                    config,
-                    quant_config=quant_config,
-                    prefix=f"{prefix}.layers.{idx}",
-                )
-                for idx in range(config.num_hidden_layers)
-            ]
-        )
+
+        from vllm.compilation.backends import set_model_tag
```
Comment (Lucaskabela) [inline vllm/model_executor/models/siglip2navit.py]: also don't need this :)

File: vllm/model_executor/models/siglip2navit.py (hunk 2 above)
Comment (tianrengao) [reply vllm/model_executor/models/siglip2navit.py:326]: I tested applying `@support_torch_compile` at the Siglip2Encoder level. Unfortunately it's incompatible with that.

`rot_pos_emb` (Line424-451) and `get_window_index` (Line453-499) both iterate over `grid_thw` values and use them to determine output tensor sizes, which causes `PendingUnbackedSymbolNotFound` in torch.compile. Keeping it at Siglip2EncoderLayer level avoids this since those computations happen in eager mode before the compiled layer forward.

File: vllm/config/utils.py [hunk 4]
```diff
@@ -296,7 +296,15 @@ def normalize_value(x):
...
-        return x.to_json_string()
+        try:
+            return x.to_json_string()
+        except (TypeError, ValueError):
+            # to_json_string() may fail for trust-remote-code configs
+            # with non-JSON-serializable nested objects. Fall back to
+            # normalizing the dict representation recursively.
+            if hasattr(x, "to_dict") and callable(x.to_dict):
+                return normalize_value(x.to_dict())
+            raise
```
Comment (ywang96) [inline vllm/config/utils.py]: Let's rebase this PR after InternVL enablement is merged.

Comment (tianrengao) [reply vllm/config/utils.py]: Just rebased this PR given InternVL is merged, thanks!

Comment (tianrengao) [reply vllm/config/utils.py]: @ywang96 would you mind add "ready" label for this pr? the previous pr has been merged https://github.com/vllm-project/vllm/pull/33827
